src/plugins/data_source/server/client/configure_client.ts

```diff
@@ -117,52 +175,75 @@ export const getCredential = async (
  * @returns child client.
  */
 const getQueryClient = async (
-  rootClient: Client,
-  dataSource: DataSourceAttributes,
+  dataSourceAttr: DataSourceAttributes,
+  addClientToPool: (endpoint: string, authType: AuthType, client: Client | LegacyClient) => void,
   cryptography?: CryptographyServiceSetup,
+  rootClient?: Client,
   requireDecryption: boolean = true
 ): Promise<Client> => {
-  const authType = dataSource.auth.type;
+  const {
+    auth: { type },
+    endpoint,
+    lastUpdatedTime,
+  } = dataSourceAttr;
 
-  switch (authType) {
+  switch (type) {
     case AuthType.NoAuth:
-      return rootClient.child();
+      addClientToPool(endpoint, type, rootClient!);
+      return rootClient!.child();
 
     case AuthType.UsernamePasswordType:
+      addClientToPool(endpoint, type, rootClient!);
       const credential = requireDecryption
-        ? await getCredential(dataSource, cryptography!)
-        : (dataSource.auth.credentials as UsernamePasswordTypedContent);
-      return getBasicAuthClient(rootClient, credential);
+        ? await getCredential(dataSourceAttr, cryptography!)
+        : (dataSourceAttr.auth.credentials as UsernamePasswordTypedContent);
+      return getBasicAuthClient(rootClient!, credential);
+
+    case AuthType.SigV4:
+      const awsCredential = requireDecryption
+        ? await getAWSCredential(dataSourceAttr, cryptography!)
+        : (dataSourceAttr.auth.credentials as SigV4Content);
+
+      const awsClient = rootClient ? rootClient : getAWSClient(awsCredential, endpoint);
+      addClientToPool(endpoint + lastUpdatedTime, type, awsClient);
+      return awsClient;
 
     default:
-      throw Error(`${authType} is not a supported auth type for data source`);
+      throw Error(`${type} is not a supported auth type for data source`);
   }
 };
 
 /**
  * Gets a root client object of the OpenSearch endpoint.
  * Will attempt to get from cache, if cache miss, create a new one and load into cache.
  *
- * @param dataSourceAttr data source saved objects attributes.
+ * @param dataSourceAttr data source saved objects attributes
  * @param config data source config
  * @returns OpenSearch client for the given data source endpoint.
  */
-const getRootClient = (
+export const getRootClient = (
```

Comment:
root client suppose to be an internal implementation, why do we need to export it?

Reply:
mistake from previous contributor. Fixed